8257020: [JVMCI] enable a JVMCICompiler to specify which GCs it supports

dougxc opened the pull request:
A number of jtreg tests require a specific GC. These tests should be ignored when EnableJVMCI is true and the JVMCI compiler does not support the required GC.

This PR adds `JVMCICompiler.isGCSupported` and makes use of it in `WhiteBox.isGCSupported`.

Prior to this PR, a test requiring a GC not yet supported by a JVMCI compiler fail as follows:
```
Error occurred during initialization of VM
JVMCI Compiler does not support selected GC: epsilon gc
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257020](https://bugs.openjdk.java.net/browse/JDK-8257020): [JVMCI] enable a JVMCICompiler to specify which GCs it supports


### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 2aa5be517355176c5417145789d70ad067415bd0


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1423/head:pull/1423`
`$ git checkout pull/1423`

vnkozlov commented:
For metropolis I have next code in jvmci_globals.cpp which switch off JVMCI if test uses unsupported GC:

```
static bool supported_gc() {
  return (UseSerialGC || UseParallelGC || UseG1GC);
}

void JVMCIGlobals::check_jvmci_supported_gc() {
  if (EnableJVMCI) {
    // Check if selected GC is supported by JVMCI and Java compiler
    if (!supported_gc()) {
      if (FLAG_IS_DEFAULT(UseJVMCINativeLibrary) && !UseJVMCINativeLibrary) { // Just disable Graal when libgraal is present
        vm_exit_during_initialization("JVMCI Compiler does not support selected GC", GCConfig::hs_err_name());
      }
      FLAG_SET_DEFAULT(EnableJVMCI, false);
      FLAG_SET_DEFAULT(UseJVMCICompiler, false);
      FLAG_SET_DEFAULT(UseJVMCINativeLibrary, false);
    }
  }
}

```

vnkozlov commented:
Why not call such `JVMCIGlobals::supported_gc()` from `GCConfig::is_gc_supported`?
I don't understand why you need to go through Java JVMCI code for that?

dougxc commented:
The reason to go through Java JVMCI code is that JVMCI itself does not know which GCs are supported by a JVMCI compiler. For example, once Graal supports ZGC, the above `supported_gc` function will be incorrect.

stefank reviewed (changes requested):
Could this be done without adding JVMCI code to GCConfig::is_gc_supported? Could you extend VMProps.java:
```
    protected void vmGC(SafeMap map) {
        var isGraalEnabled = Compiler.isGraalEnabled();
        for (GC gc: GC.values()) {
            map.put("vm.gc." + gc.name(),
                    () -> "" + (gc.isSupported()
                            && (!isGraalEnabled || isGcSupportedByGraal(gc))
                            && (gc.isSelected() || GC.isSelectedErgonomically())));
        }
    }
```
and add:
`!isJvcmiEnabled || isGcSupportedByJvmci(gc)`
plus necessary JVMCI code to implement that?

dougxc commented:
> Could this be done without adding JVMCI code to GCConfig::is_gc_supported?

This has to go through the WhiteBox API as JVMCI is not accessible to normal Java code (such as VMProps).
We could restrict the JVMCI interposition to WhiteBox. This is what I did [initially](https://github.com/openjdk/jdk/pull/1423/commits/191d61d996dd08f43c2c946f02b0cce0d96f2ef4). However, I then saw that `GCConfig::is_gc_supported` is only called from whitebox.cpp so [moved](https://github.com/openjdk/jdk/pull/1423/commits/54b7ba8463dfee9a679abb09a5cc898ba8550d85) the logic directly into `GCConfig`. I think this is better because as soon as there's another caller of `GCConfig::is_gc_supported`, they should get an accurate answer and not have to worry about implications of `EnableJVMCI` separately.

stefank commented:
> > Could this be done without adding JVMCI code to GCConfig::is_gc_supported?
> 
> This has to go through the WhiteBox API as JVMCI is not accessible to normal Java code (such as VMProps).
> We could restrict the JVMCI interposition to WhiteBox. This is what I did [initially](https://github.com/openjdk/jdk/pull/1423/commits/191d61d996dd08f43c2c946f02b0cce0d96f2ef4). However, I then saw that `GCConfig::is_gc_supported` is only called from whitebox.cpp so [moved](https://github.com/openjdk/jdk/pull/1423/commits/54b7ba8463dfee9a679abb09a5cc898ba8550d85) the logic directly into `GCConfig`. I think this is better because as soon as there's another caller of `GCConfig::is_gc_supported`, they should get an accurate answer and not have to worry about implications of `EnableJVMCI` separately.

We already have code in place to ensure that EnableJVMCI is set to false when you run a GC that doesn't support JVMCI:
```
void JVMCIGlobals::check_jvmci_supported_gc() {
  if (EnableJVMCI) {
    // Check if selected GC is supported by JVMCI and Java compiler
    if (!(UseSerialGC || UseParallelGC || UseG1GC)) {
      vm_exit_during_initialization("JVMCI Compiler does not support selected GC", GCConfig::hs_err_name());
      FLAG_SET_DEFAULT(EnableJVMCI, false);
      FLAG_SET_DEFAULT(UseJVMCICompiler, false);
    }
  }
}
```

So, in our view we have "if you run this particular GC, that don't have support in JVMCI, then we turn off JVMCI forcefully", and we rely on that behavior in the HotSpot code. The proposed patch turns this inside out and the changed is_gc_supported function now becomes "if you've turned on EnableJVMCI is this GC supported". It sort-of inverts the dependency between the two sub-systems and we don't want to use that particular question in HotSpot. AFAICT, this is something you only need to ask in the context of jtreg's @requires, so I think it should be moved out of the GC code and put in the jtreg extension code.

We've other flags that we treat similarly. For example, UseCompressedOops. ZGC doesn't support it, so we turn it off forcefully, but we don't have code in GCConfig::is_gc_supported that returns false when UseCompressedOops is turned on. That is completely handled by the jtreg extensions (and the WhiteboxAPI).

I think you could easily solve this without adding calls through JVMCI from the GC code by:
1) Update JVMCIGlobals::supported_gc() to make a call through JVMCI into the used compiler that can tell if it supports the particular GC
2) Use Vladimirs proposed check_jvmci_supported_gc function, which uses supported_gc().
3) Add your own  WhiteboxAPI function, say WB_IsJVMCISupportedByGC, and call your "supported_gc" function from that
4) Update VMProps.java like I suggested above

dougxc commented:
`JVMCIGlobals::check_jvmci_supported_gc` is incorrect in that it hard codes the GCs supported by JVMCI which not something JVMCI itself can accurately answer. What's more, the upcall into JVMCI Java code to get the right answer cannot be made here as it is too early in the VM boot process. There's not even a `JavaThread` available yet. This code should be removed and let Graal exit the VM with an error should it be used in conjunction with a GC it does not support. For example:
```
Error occurred during initialization of VM
JVMCI Compiler does not support selected GC: epsilon gc
```

Moving the check to later in the VM bootstrap doesn't really work either as it imposes the overhead of initializing JVMCI and Graal eagerly.
This means a conflict between GC and JVMCI compiler cannot be finessed by overriding the compiler selection to use a non-JVMCI compiler. It's a fatal VM error, just like selecting multiple GCs on the command line is.

If you agree with that, then I think the proposed PR is the right approach. We could revert to the version that leaves the JVMCI specific logic in `WB_IsGCSupported` but as I stated, that means `GCConfig::is_gc_supported` can give a wrong answer.

vnkozlov commented:
I think you merged 2 things into this PR which made me confused. The title says that Graal should specify which GC it supports (new JVMCI API). But your description is about fixing Graal testing with different GCs. Yes, they are connected but I think they should be reviewed separately.

The reason currently we test Graal supported GC in C++ code during VM startup is because we want to avoid executing application in Interpreter and bailout later when Graal is loaded to compile a hot method. Note, Graal in JDK is loaded on first compilation request.
To get answer from Graal about GC you would have to load and initialize it very early which will affect startup (even with libgraal).

To add new JVMCI API just to resolve testing issues is overkill for me.

dougxc commented:
Now I'm confused ;-) The title and description of this PR don't mention Graal at all. While Graal is obviously the concrete JVMCI compiler motivating this change, the problem would relate to any Java based JVMCI compiler. This is why an API change is needed.

I understand the motivation for wanting to test which GCs are supported by JVMCI in C++ code during VM startup. But I don't think that can justify giving a wrong answer. What happens when support for ZGC is added to Graal? The VM will incorrectly exit during startup based on the current code.

With JVMCI, the problem of delayed VM exit upon misconfiguration already exists. For example, an incorrectly specified or unrecognized `-Dgraal.*` option will only be detected when compiling the first method with Graal (i.e. when Graal is lazily initialized). Maybe I'm missing something, but why is this a real problem? Does it matter whether the VM exits after 1ms of execution or 500ms?

vnkozlov commented:
Actually here is fundamental question. Why not build GraalVM JDK without GCs which Graal does not support? It is all configurable.

dougxc commented:
> Actually here is fundamental question. Why not build GraalVM JDK without GCs which Graal does not support? It is all configurable.

GraalVM could indeed do that but there are other OpenJDK community members who want to be able to use Graal on stock JDK binaries. Even for GraalVM, we want to deviate as little as possible from how the JDK underlying GraalVM is built.

vnkozlov commented:
Yes, it is reasonable case. We already had such situation with CMS before. You and they understand which GC can be used with Graal. I agree with removal of `JVMCIGlobals::check_jvmci_supported_gc()` and not do any C++ checks during VM startup.

Now about testing and changes. I agree with @stefank that you should not modify `GCConfig::is_gc_supported()` because it is about Hotspot VM support (`Interpreter` only, for example).  I suggest to modify `isGcSupportedByGraal()` in VMProps.java by adding new `WB_IsGCSupportedByGraal()` WB api to call JVMCI runtime. Then `vmGC()` in VMProps.java will work as it is. And you can change `Graal` with `JVMCICompiler` in methods names if you want.

dougxc commented:
Ok, I get the point about `GCConfig::is_gc_supported()` now. It's answering the question "is the given GC built into the VM" as opposed to "can the given GC work in the current VM runtime configuration".

Based on your suggestions, I've removed the JVMCI logic from `GCConfig`, added `WB_isGCSupportedByJVMCI` (it's better to keep things JVMCI compiler neutral at this layer) and modified `VMProps` to use it (https://github.com/openjdk/jdk/pull/1423/commits/379f48f794f2f59e1b4ffd837025f9ea999d06ff).

I've also implemented the Graal side of this interface so that it can be demonstrated to work in the JDK (https://github.com/openjdk/jdk/pull/1423/commits/384287b748acfbb3ebea7d9b5eb1cbbe106d912d):
```
> sh configure --disable-precompiled-headers --enable-jvm-feature-graal
> make images

# With Graal
> jtreg -v -noreport -jdk:build/macosx-x86_64-server-release/images/jdk -vmoptions:"-XX:+UnlockExperimentalVMOptions -XX:+UseJVMCICompiler" test/hotspot/jtreg/gc/epsilon/TestAlignment.java
Test results: no tests selected
Results written to /Users/dnsimon/jdk-jdk/open/JTwork

# Without Graal
> jtreg -v -noreport -jdk:build/macosx-x86_64-server-release/images/jdk test/hotspot/jtreg/gc/epsilon/TestAlignment.java
runner starting test: gc/epsilon/TestAlignment.java
runner finished test: gc/epsilon/TestAlignment.java
Passed. Execution successful
Test results: passed: 1
Results written to /Users/dnsimon/jdk-jdk/open/JTwork
```

vnkozlov reviewed (approved):
Looks good.

dougxc commented:
Thanks for the review @vnkozlov .
Are you also good with the PR is its current state @stefank ?

vnkozlov commented on src/hotspot/share/jvmci/jvmci_globals.cpp:
```diff
@@ -204,7 +204,7 @@ void JVMCIGlobals::check_jvmci_supported_gc() {
     // Check if selected GC is supported by JVMCI and Java compiler
     if (!(UseSerialGC || UseParallelGC || UseG1GC)) {
-      vm_exit_during_initialization("JVMCI Compiler does not support selected GC", GCConfig::hs_err_name());
+      log_warning(gc)("Selected GC does not support JVMCI: %s", GCConfig::hs_err_name());
```
I would add new UL tag `jvmci` if we don't have it yet and use it here. Also message should say `EnableJVMCI is turned off because ...'

vnkozlov reviewed (commented):
Latest changes looks like reversal of original. Will GraalVM have additional checks to allow ask Graal before disabling JVMCI?

dougxc commented:
> Latest changes looks like reversal of original. Will GraalVM have additional checks to allow ask Graal before disabling JVMCI?

No, the changes in this PR are sufficient. When JVMCI support is added for Graal to be able to implement support for ZGC/Shenandoah/Epsilon, the `check_jvmci_supported_gc` implementation will be updated.

vnkozlov commented:
> > Latest changes looks like reversal of original. Will GraalVM have additional checks to allow ask Graal before disabling JVMCI?
> 
> No, the changes in this PR are sufficient. When JVMCI support is added for Graal to be able to implement support for ZGC/Shenandoah/Epsilon, the `check_jvmci_supported_gc` implementation will be updated.

Okay. If you are fine with that.

dougxc commented:
There is the interesting question of whether explicit JVMCI API extensions are needed for Epsilon. @shipilev what's your intuition on that?

vnkozlov commented on src/hotspot/share/jvmci/jvmci_globals.cpp:
```diff
@@ -204,7 +204,7 @@ void JVMCIGlobals::check_jvmci_supported_gc() {
     // Check if selected GC is supported by JVMCI and Java compiler
     if (!(UseSerialGC || UseParallelGC || UseG1GC)) {
-      vm_exit_during_initialization("JVMCI Compiler does not support selected GC", GCConfig::hs_err_name());
+      log_warning(gc, jvmci)("Setting EnableJVMCI to false as selected GC does not support JVMCI: %s", GCConfig::hs_err_name());
```
Now message is:
```
[0.001s][warning][gc,jvmci] Setting EnableJVMCI to false as selected GC does not support JVMCI: z gc
```

Okay.

stefank reviewed (approved):
I've been talking to Doug the last few days and we've settled on this approach that both solves their testing needs, and at the same time satisfy my wish to *not* turn on EnableJVMCI for GCs that don't have fully support for it.

dougxc commented:
Thanks a lot for the diligent review and helpful contributions @stefank !

dougxc commented:
/integrate
